hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py

```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+    num_selected = selected.sum()
+    num_tiles = len(image_tiles)
+    num_discarded = num_tiles - num_selected
+    percentage_discarded = 100 * num_discarded / num_tiles
+    logging.info(f"Discarded {num_discarded}/{num_tiles} tiles ({percentage_discarded:.2f} %)")
 
     image_tiles = image_tiles[selected]
     mask_tiles = mask_tiles[selected]
     tile_locations = tile_locations[selected]
     occupancies = occupancies[selected]
 
-    abs_tile_locations = (sample['scale'] * tile_locations + sample['location']).astype(int)
+    top, left = sample['location']
+    offset = np.array((left, top))
+
+    abs_tile_locations = (sample['scale'] * tile_locations + offset).astype(int)
+    tile_size_scaled = int(tile_size * sample['scale'])
+    tile_boxes = [Box(x, y, tile_size_scaled, tile_size_scaled) for x, y in abs_tile_locations]
 
-    return image_tiles, mask_tiles, abs_tile_locations, occupancies, n_discarded
+    return image_tiles, mask_tiles, tile_boxes, occupancies, num_discarded
 
 
 # TODO refactor this to separate metadata identification from saving. We might want the metadata
 # even if the saving fails
 def save_tile(sample: dict, image_tile: np.ndarray, mask_tile: np.ndarray,
-              tile_location: Sequence[int], output_dir: Path) -> dict:
-    slide_id = sample['image_id']
-    descriptor = get_tile_descriptor(tile_location)
-    image_tile_filename = f"train_images/{descriptor}.png"
-    mask_tile_filename = f"train_label_masks/{descriptor}_mask.png"
+              tile_box: Box, output_dir: Path) -> dict:
+    slide_id = sample[SlideKey.SLIDE_ID]
+    tile_id = get_tile_id(slide_id, tile_box)
+    image_tile_filename = f"train_images/{tile_id}.png"
+    mask_tile_filename = f"train_label_masks/{tile_id}_mask.png"
 
     save_image(image_tile, output_dir / image_tile_filename)
     save_image(mask_tile, output_dir / mask_tile_filename)
 
+    slide_metadata = sample['metadata']
     tile_metadata = {
-        'slide_id': slide_id,
-        'tile_id': get_tile_id(slide_id, tile_location),
-        'image': image_tile_filename,
-        'mask': mask_tile_filename,
-        'tile_x': tile_location[0],
-        'tile_y': tile_location[1],
-        'data_provider': sample['data_provider'],
-        'slide_isup_grade': sample['isup_grade'],
-        'slide_gleason_score': sample['gleason_score'],
+        TileKey.SLIDE_ID.value: slide_id,
+        TileKey.TILE_ID.value: tile_id,
+        TileKey.IMAGE.value: image_tile_filename,
+        TileKey.MASK.value: mask_tile_filename,
+        TileKey.TILE_LEFT.value: tile_box.x,
+        TileKey.TILE_TOP.value: tile_box.y,
+        TileKey.TILE_RIGHT.value: tile_box.x + tile_box.w,
+        TileKey.TILE_BOTTOM.value: tile_box.y + tile_box.h,
+        'data_provider': slide_metadata['data_provider'],
+        'slide_isup_grade': slide_metadata['isup_grade'],
+        'slide_gleason_score': slide_metadata['gleason_score'],
```

Comment:
Couldn't find Enums for these.

Reply:
Correct; metadata fields are free-form and specific to each dataset. That said, they're listed in `PandaDataset.METADATA_COLUMNS`. Not sure it's worth using that here, though.